src/dbnode/storage/series/buffer.go

```diff
@@ -89,53 +99,50 @@ type databaseBuffer interface {
 
 	Stats() bufferStats
 
-	// MinMax returns the minimum and maximum blockstarts for the buckets
-	// that are contained within the buffer. These ranges exclude buckets
-	// that have already been drained (as those buckets are no longer in use.)
-	MinMax() (time.Time, time.Time, error)
-
-	Tick() bufferTickResult
-
-	NeedsDrain() bool
+	Tick(versions map[xtime.UnixNano]BlockState) bufferTickResult
 
-	DrainAndReset() drainAndResetResult
-
-	Bootstrap(bl block.DatabaseBlock) error
+	Bootstrap(bl block.DatabaseBlock)
 
 	Reset(opts Options)
 }
 
 type bufferStats struct {
-	openBlocks  int
 	wiredBlocks int
 }
 
-type drainAndResetResult struct {
-	mergedOutOfOrderBlocks int
-}
-
 type bufferTickResult struct {
 	mergedOutOfOrderBlocks int
+	evictedBuckets         int
 }
 
 type dbBuffer struct {
-	opts              Options
-	nowFn             clock.NowFn
-	drainFn           databaseBufferDrainFn
-	pastMostBucketIdx int
-	buckets           [bucketsLen]dbBufferBucket
-	blockSize         time.Duration
-	bufferPast        time.Duration
-	bufferFuture      time.Duration
+	opts  Options
+	nowFn clock.NowFn
+
+	bucketsMap map[xtime.UnixNano]*BufferBucketVersions
+	// Cache of buckets to avoid map lookup of above.
+	bucketVersionsCache [bucketsCacheSize]*BufferBucketVersions
```

Comment:
For the vast majority of the case, we'll be constantly accessing the map for the current bucket (for real time metrics). After benchmarking, we saw that accessing maps was significantly more expensive than going through an array, so we cache the pointers of buckets that we will most likely be using.

The array size is 2 because at a block boundary, warm writes can be for either block.